backend-lib/src/main/rust/lib.rs

```diff
@@ -1148,23 +1016,245 @@ pub unsafe extern "C" fn Java_cash_z_ecc_android_sdk_internal_jni_RustBackend_re
+        };
+
+        db_data
+            .put_sapling_subtree_roots(start_index, &roots)
+            .map(|()| JNI_TRUE)
+            .map_err(|e| format_err!("Error while storing Sapling subtree roots: {}", e))
+    });
+
+    unwrap_exc_or(&env, res, JNI_FALSE)
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn Java_cash_z_ecc_android_sdk_internal_jni_RustBackend_updateChainTip(
+    env: JNIEnv<'_>,
+    _: JClass<'_>,
+    db_data: JString<'_>,
+    height: jlong,
+    network_id: jint,
+) -> jboolean {
+    let res = panic::catch_unwind(|| {
+        let network = parse_network(network_id as u32)?;
+        let mut db_data = wallet_db(&env, network, db_data)?;
+        let height = BlockHeight::try_from(height)?;
+
+        db_data
+            .update_chain_tip(height)
+            .map(|()| JNI_TRUE)
+            .map_err(|e| format_err!("Error while updating chain tip to height {}: {}", height, e))
+    });
+
+    unwrap_exc_or(&env, res, JNI_FALSE)
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn Java_cash_z_ecc_android_sdk_internal_jni_RustBackend_getFullyScannedHeight(
+    env: JNIEnv<'_>,
+    _: JClass<'_>,
+    db_data: JString<'_>,
+    network_id: jint,
+) -> jlong {
+    let res = panic::catch_unwind(|| {
+        let network = parse_network(network_id as u32)?;
+        let db_data = wallet_db(&env, network, db_data)?;
+
+        match db_data.block_fully_scanned() {
+            Ok(Some(metadata)) => Ok(i64::from(u32::from(metadata.block_height()))),
+            // Use -1 to return null across the FFI.
+            Ok(None) => Ok(-1),
+            Err(e) => Err(format_err!(
+                "Failed to read block metadata from WalletDb: {:?}",
+                e
+            )),
+        }
+    });
+    unwrap_exc_or(&env, res, -1)
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn Java_cash_z_ecc_android_sdk_internal_jni_RustBackend_getMaxScannedHeight(
+    env: JNIEnv<'_>,
+    _: JClass<'_>,
+    db_data: JString<'_>,
+    network_id: jint,
+) -> jlong {
+    let res = panic::catch_unwind(|| {
+        let network = parse_network(network_id as u32)?;
+        let db_data = wallet_db(&env, network, db_data)?;
+
+        match db_data.block_max_scanned() {
+            Ok(Some(metadata)) => Ok(i64::from(u32::from(metadata.block_height()))),
+            // Use -1 to return null across the FFI.
+            Ok(None) => Ok(-1),
+            Err(e) => Err(format_err!(
+                "Failed to read block metadata from WalletDb: {:?}",
+                e
+            )),
+        }
+    });
+    unwrap_exc_or(&env, res, -1)
+}
+
+fn encode_scan_progress(env: &JNIEnv<'_>, progress: Ratio<u64>) -> Result<jobject, failure::Error> {
+    let output = env.new_object(
+        "cash/z/ecc/android/sdk/internal/model/JniScanProgress",
+        "(JJ)V",
+        &[
+            JValue::Long(*progress.numerator() as i64),
+            JValue::Long(*progress.denominator() as i64),
+        ],
+    )?;
```

Comment:
I don't know what this will do if the constructor throws an `IllegalArgumentException` (which is what Kotlin's [`require`](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin/require.html) does if it fails). It's not documented by [JNIEnv::new_object](https://docs.rs/jni/latest/jni/struct.JNIEnv.html#method.new_object). I assume you would have to call [`JNIEnv::exception_occurred`](https://docs.rs/jni/latest/jni/struct.JNIEnv.html#method.exception_occurred), [`exception_clear`](https://docs.rs/jni/latest/jni/struct.JNIEnv.html#method.exception_clear), etc.

Reply:
Oh, what happens is that on exit from a JNI native method, the JVM checks whether there is a "pending" exception and re-throws it. So if `encode_scan_progress` is used from another JNI native method, the `IllegalArgumentException` will be re-thrown. However, in order for the Rust code not to do the wrong thing in the meantime, `JNIEnv::new_object` would have to check for a pending exception and returns an `Err` if there is one. Which it does, but I had to check the [implementations](https://github.com/jni-rs/jni-rs/blob/v0.20.0/src/wrapper/macros.rs) of the jni-rs macros to confirm that.